### PR TITLE
Add typing to ``PyLinter.reporter``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |

## Description

We always expect `reporter` not to be `None`, but we do not initialise it as such. The comment in `load_default_plugins` shows that it is often necessary to have one.
Furthermore, this makes the typing of `PyLinter.reporter` much easier as we won't need to do `if linter.reporter` anymore.

I will add a changelog entry for `2.13` once `2.12` is out. Therefore currently blocked.

### CODE REVIEW

File: tests/test_self.py [hunk 1]
```diff
@@ -1259,3 +1259,10 @@ def test_enable_all_extensions() -> None:
             exit=False,
         )
         assert sorted(plugins) == sorted(runner.linter._dynamic_plugins)
+
+    @staticmethod
+    def test_load_text_repoter_if_not_provided() -> None:
+        """Test if PyLinter.reporter is a TextRepoter is no reporter is provided"""
```
Comment (Pierre-Sassoulas) [inline tests/test_self.py:1265]:
```suggestion
        """Test if PyLinter.reporter is a TextReporter if no reporter is provided"""
```

File: pylint/lint/pylinter.py [hunk 2]
```diff
@@ -567,16 +580,10 @@ def __init__(self, options=(), reporter=None, option_groups=(), pylintrc=None):
...
-        if reporter:
-            self.set_reporter(reporter)
 
     def load_default_plugins(self):
         checkers.initialize(self)
         reporters.initialize(self)
-        # Make sure to load the default reporter, because
-        # the option has been set before the plugins had been loaded.
-        if not self.reporter:
-            self._load_reporters()
```
Comment (Pierre-Sassoulas) [inline pylint/lint/pylinter.py:579]: I like how it already simplify the code by removing handling that is dispersed in place it should not be.

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM !

File: pylint/lint/pylinter.py (hunk 2 above)
Comment (DanielNoord) [reply pylint/lint/pylinter.py:579]: `MultiReporter` might create some additional problems for us. Have not looked into it fully, but having a reporter "that's not actually a reporter" on `PyLinter.reporter` might be troublesome..

**Comment by areveny** [comment]
This looks like a good change to make the reporter explicit. I think the ``Run`` wrapper introduces a text reporter through the default options, but it's not intuitive how ith appens.

Since I was just looking at the reporters, I tried out some of the command line arguments I was testing with and they have the same functionality as before with this change.

**Pull request merged**